lib/assert.js

```diff
@@ -174,10 +174,16 @@ function wrapAssertions(callbacks) {
 			if (promise) {
 				// Record stack before it gets lost in the promise chain.
 				const stack = getStack();
-				const intermediate = promise.then(makeNoop, makeRethrow).then(fn => test(fn, stack));
+				const intermediate = promise.then(value => {
+					throw new AssertionError({
+						assertion,
+						message: 'Expected promise to be rejected, but it was resolved instead',
+						values: [formatAssertError.formatWithLabel('Resolved with:', value)]
+					});
+				}, reason => test(makeRethrow(reason), stack));
+
 				pending(this, intermediate);
-				// Don't reject the returned promise, even if the assertion fails.
-				return intermediate.catch(noop);
+				return intermediate;
```

Comment:
The `catch` is necessary. We don't expose the `AssertionError` instance to the test author. Instead add a `.catch()` callback in the `pending()` implementation here: <https://github.com/tdeschryver/ava/blob/4a19a96da89a1533025295a7cd78b74c12528f04/test/assert.js#L15>.